xayn-ai/src/coi/config.rs

```diff
@@ -1,100 +1,172 @@
 use std::{cmp::Ordering, time::Duration};
 
-use crate::embedding::utils::COSINE_SIMILARITY_RANGE;
 use displaydoc::Display;
 use thiserror::Error;
 
-use crate::utils::{nan_safe_f32_cmp_desc, SECONDS_PER_DAY};
+use crate::{
+    embedding::utils::COSINE_SIMILARITY_RANGE,
+    utils::{nan_safe_f32_cmp_desc, SECONDS_PER_DAY},
+};
 
-/// The configuration of the ranker.
+/// The configuration of the cois.
 #[derive(Clone, Debug)]
-pub struct Config {
+struct CoiConfig {
     shift_factor: f32,
     threshold: f32,
+    min_positive_cois: usize,
+    min_negative_cois: usize,
+}
+
+impl Default for CoiConfig {
+    fn default() -> Self {
+        Self {
+            shift_factor: 0.1,
+            threshold: 0.67,
+            min_positive_cois: 2,
+            min_negative_cois: 2,
+        }
+    }
+}
+
+/// The configuration of the kpe.
+#[derive(Clone, Debug)]
+struct KPEConfig {
     horizon: Duration,
     gamma: f32,
     penalty: Vec<f32>,
-    min_positive_cois: usize,
-    min_negative_cois: usize,
 }
 
-/// Potential errors of the ranker configuration.
+impl Default for KPEConfig {
+    fn default() -> Self {
+        Self {
+            horizon: Duration::from_secs(SECONDS_PER_DAY as u64 * 30),
+            gamma: 0.9,
+            penalty: vec![1., 0.75, 0.66],
+        }
+    }
+}
+
+/// The configuration of the coi system.
+#[derive(Clone, Debug, Default)]
+pub struct Config {
+    coi: CoiConfig,
+    kpe: KPEConfig,
+}
+
+/// Errors of the coi system configuration.
 #[derive(Copy, Clone, Debug, Display, Error)]
 pub enum Error {
     /// Invalid coi shift factor, expected value from the unit interval
     ShiftFactor,
     /// Invalid coi threshold, expected non-negative value
     Threshold,
-    /// Invalid coi gamma, expected value from the unit interval
-    Gamma,
-    /// Invalid coi penalty, expected non-empty, finite and sorted values
-    Penalty,
     /// Invalid minimum number of positive cois, expected positive value
     MinPositiveCois,
     /// Invalid minimum number of negative cois, expected positive value
     MinNegativeCois,
+    /// Invalid coi gamma, expected value from the unit interval
+    Gamma,
+    /// Invalid coi penalty, expected non-empty, finite and sorted values
+    Penalty,
 }
 
 impl Config {
-    /// The shift factor by how much a Coi is shifted towards a new point.
+    /// The shift factor by how much a coi is shifted towards a new point.
     pub fn shift_factor(&self) -> f32 {
-        self.shift_factor
+        self.coi.shift_factor
     }
 
     /// Sets the shift factor.
     ///
     /// # Errors
     /// Fails if the shift factor is outside of the unit interval.
-    pub fn with_shift_factor(self, shift_factor: f32) -> Result<Self, Error> {
+    pub fn with_shift_factor(mut self, shift_factor: f32) -> Result<Self, Error> {
         if (0. ..=1.).contains(&shift_factor) {
-            Ok(Self {
```

Comment:
it's just for readability due to the new nesting, the field-changing logic stays the same, otherwise we would need to write sth like
```rust
Ok(Self {
    coi: CoiConfig {
        shift_factor,
        ..self.coi
    },
    ..self
})
```